feat: handle optional fields in bridge (#54)
openedx/openedx-events: docs/decisions/0007-optional-fields.rst

the field if the user does not supply one at instantiation.

`The Avro documentation <https://avro.apache.org/docs/current/spec.html#schemas>`_ describes somewhat similar mechanics for its fields. If a field's ``type`` list includes
"null," the user can set the value of the field to null/None. Separately, if a field is defined with a ``default``
parameter, then the user does not have to set the value at all. However, it seems the ``default`` parameter may not be
fully-supported with Kafka (see https://github.com/confluentinc/kafka-rest/issues/427 ). The Avro documentation,
in contrast to that of ``attrs``, uses the default parameter to define optionality.

Decision
--------

- The bridge will convert ``attrs`` fields with a default value of None to Avro fields with type = ["null",<type>] and default = "null".

Consequences
------------

- The notions of "optionality" and "nullability" will be effectively conflated. This means no distinctions will be made between fields for which ``None`` is a valid value and fields that simply do not need to be supplied.
- Attr fields with defaults that are not ``None`` will be treated as required

> - At the point of writing this ADR, it is unclear exactly how Kafka will treat the addition and removal of optional fields for the purposes of schema evolution (because of the issue linked above). This will require a little more testing.

Reviewer: I’m not sure if these notes are useful for this ADR, but I’ll let you decide. I know we still need to prove out schema evolution doing what we need it to.

Deferred/Rejected Decisions
---------------------------

> - We are deferring any work to around making top-level fields in event data optional. This reverses the plan laid out in ":doc:`0006-event-schema-serialization-and-evolution`", which stated that the initial bridge would treat all new top-level fields as optional and deferred any work around making sub-fields optional

Reviewer: 1. I don't think ADR 6 means to say what you are reading. Maybe we need to fix some wording there? I adjusted the text in this ADR as I think it should read.
2. Fixed typo "work to around".
```suggestion
- Top-level fields will continue to be required, as stated in ":doc:`0006-event-schema-serialization-and-evolution`". We continue to defer the possibility of making these optional.
```

===========================================

Status
------

Provisional

Context
-------
The ``attrs`` documentation specifically defines the notion of optionality as, "An optional attribute is one which
can be set to None in addition to satisfying the requirements of the sub-validator." Importantly, under this definition,
the user must still provide a value for this optional field when instantiating the ``attrs``-decorated class,
but this value can be ``None``. Separately, attrs fields can have a ``default`` value, which will provide a value for
the field if the user does not supply one at instantiation.

The Avro documentation describes somewhat similar mechanics for its fields. If a field's ``type`` list includes
"null," the user can set the value of the field to null/None. Separately, if a field is defined with a ``default``
parameter, then the user does not have to set the value at all. However, it seems the ``default`` parameter may not be
fully-supported with Kafka (see https://github.com/confluentinc/kafka-rest/issues/427 ). The Avro documentation,
similar to that of ``attrs``, uses the default parameter to define optionality.

Decision
--------

> - The bridge will convert ``attrs`` fields with a default value of None to Avro fields with type = ["null",<type>]

Author: Yes, didn't want to contemplate that for this particular issue

and default = "null".

> - Top-level fields in event data definitions will not be optional. This is contrary to and overrides the decision made in

Reviewer: I need to look more closely at this. I don’t think we are overriding a decision around the top-level fields. I think we simply haven’t implemented any capability around that, which was also true before.

Author: there were some capabilities around it before the previous PR that changed the bridge to work with entire signals, but they were removed in favor of getting the basics of the bridge out faster.
